### ISSUE TITLE
[radm] Reproduce RADM training with a genuine package model and Detectron2 reference adapter

### ISSUE THREAD

**Issue by creative-graphic-design-dev** [creative-graphic-design/design-generators#261]
# [radm] Reproduce RADM training with a genuine package model and Detectron2 reference adapter

Create a separate future PR for genuine RADM training reproduction. This work is intentionally excluded from PR #209, which remains the inference/conversion/reference scaffold defined by issue #35.

References: #35, #60, #148, #212, #249, and #253.

## Scope

Implement package-local PyTorch Lightning training for RADM and establish staged training parity against the original Detectron2 implementation.

The package-side training model must be the actual runtime RADM model used by inference and checkpoint conversion. A second simplified `RADMTrainableModel` is not acceptable.

The original implementation is [JD-GenX/RADM](https://github.com/JD-GenX/RADM), especially `RADM/detector.py`, `RADM/head.py`, `RADM/dataset_mapper.py`, `RADM/config.py`, `train_net.py`, `configs/Base-RADM.yaml`, and `configs/radm.yaml`.

## Original topology to reproduce

The reference adapter must construct the original Detectron2 graph rather than injecting a vendor model into a package wrapper.

The active topology includes:

- Detectron2 `RADM` meta-architecture registered through `META_ARCH_REGISTRY`.
- ResNet-50 FPN backbone with `res2`–`res5` outputs consumed as `p2`–`p5`.
- Detectron2 `ROIAlignV2`, resolution `7`, sampling ratio `2`.
- 100 proposal boxes and hidden dimension `256`.
- Six repeated `RCNNHead` blocks in `DynamicHead`.
- Proposal self-attention, `DynamicConv`, visual-textual relation, and geometric relation modules.
- 768-dimensional text features, padded to 20 entries with a text mask.
- 1,000 diffusion timesteps and `SNR_SCALE=2.0`.
- Diffusion target preparation, noise/timestep sampling, denoising heads, auxiliary deep-supervision outputs, and final loss computation.
- Focal classification loss plus L1 and generalized-IoU box losses, including original matching and weighting behavior.

The original training orchestrator is a custom `Trainer(DefaultTrainer)` that constructs the model, optimizer, loader, DDP/AMP/SimpleTrainer, LR scheduler, EMA/evaluation/checkpoint hooks, and then executes the training loop. The effective config includes AdamW, learning rate `2.5e-5`, weight decay `1e-4`, milestones `150000` and `220000`, maximum iteration `250000`, warmup factor `0.01` for `1000` iterations, full-model gradient clipping with norm `1.0`, and the original loader/evaluation settings.

The S0 topology guard must explicitly expose the original discrepancy between the default four-class RADM config and the five-class CGL vocabulary. No class-count correction may be silently introduced.

## Package model and reference adapter

Add or revise the package training surface under:

- `models/radm/src/radm/training/`
- `models/radm/configs/training/`
- `models/radm/tests/`
- `models/radm/tests/vendor_parity/`

Required components include:

- `RADMTrainingModule`, wrapping the runtime `RADMDenoiser`.
- `RADMDataModule`.
- A dataset adapter and collator matching effective Detectron2 mapper behavior.
- Model-local loss and trace helpers only where required by the runtime model.
- A vendor-only `RADMReferenceAdapter` under gated parity tooling.

The reference adapter must load the original Detectron2 model with the `vendor` extra, apply the original `CfgNode` configuration and mapper, expose original forward/loss/optimizer/scheduler/loader behavior at trace points, remain isolated from normal package imports, and never replace the package model in package-side S0–S4 checks.

The package model must have an explicit checkpoint/state-dict mapping to the original topology. Missing keys must fail; extra keys require an explicit allowlist and explanation.

## Lazy-import dependency

The future RADM training surface must adopt the shared lazy-import contract from [issue #260](https://github.com/creative-graphic-design/design-generators/issues/260). That contract must be finalized before the RADM training PR merges. RADM training must not define a package-local competing convention.

Required behavior:

- Importing `radm` for inference must not import Lightning or Detectron2.
- `radm.training` must not eagerly import Lightning at package import time.
- The `training` extra and `traingen` entry point may import Lightning only when training is explicitly invoked.
- Vendor/reference imports remain behind gated vendor-parity paths.
- `models/radm/pyproject.toml`, `radm/__init__.py`, and `radm/training/__init__.py` follow the shared contract.

This dependency must not be implemented or coupled into PR #209.

## LightningCLI and YAML surface

Use the shared `traingen` and `traingen-parity` workspace packages and the repository training protocol. Configs must use LightningCLI `class_path` and `init_args`; do not use Hydra `_target_`, `defaults`, interpolation, or `hydra.*` configuration.

Provide separate configs for vendor-compatible CGL training, vendor-compatible CGL-v2 training, strict deterministic parity, and small smoke/configuration checks.

```bash
uv run --package radm --extra training traingen fit \
  --config models/radm/configs/training/<config>.yaml
```

Preserve one inference/training state-dict surface so trained checkpoints can be converted and loaded through the existing RADM pipeline.

## Optimizer, scheduler, and loader parity

Reproduce effective original values and update order:

- AdamW, base learning rate `2.5e-5`, and weight decay `1e-4`.
- Batch size `16` unless a documented parity scaling rule applies.
- Full-model gradient clipping with norm `1.0`.
- Warmup factor `0.01` for `1000` iterations.
- LR milestones `150000` and `220000`; maximum iteration count `250000`.
- Step-based scheduler cadence, with Lightning configured explicitly using `interval="step"`.
- Detectron2 optimizer parameter grouping and backbone multiplier behavior.
- Original AMP/DDP/SimpleTrainer branch selection.
- Checkpoint and evaluation hook ordering, including EMA if enabled.
- Original `NUM_WORKERS=0` and loader ordering for strict deterministic evidence.

S2/S3 evidence must verify scheduler state against `trainer.global_step`.

## CGL and CGL-v2 data boundaries

Use the canonical sources:

- `creative-graphic-design/CGL-Dataset`
- `creative-graphic-design/CGL-Dataset-v2`

Preparation is limited to making original training inputs reproducible:

- Preserve five-class source order and record source labels plus package-facing mapping.
- Materialize or stream the COCO-style structure expected by the original loader.
- Preserve image/annotation ids, dimensions, pixel boxes, category ids, and split membership.
- Reproduce `layout_train`/`layout_val` registration, random flip, `ResizeShortestEdge`, optional crop behavior, image normalization, and annotation transforms.
- Preserve original train/test text-feature directories, 768-dimensional tensors, and 20-entry masks.
- Do not invent a text encoder or claim text-feature generation parity.
- Missing text features must either reproduce the original fallback as an explicitly tested branch or block the claimed run; never silently substitute zeros.
- Use CGL-v2 `ralf-style` when validation, no-annotation, or saliency evaluation is required; the default config has no validation split.
- Do not use Desigen as a CGL/Crello substitute.
- Do not download full datasets in ordinary tests.
- Store staged data, checkpoints, and evidence under durable paths such as `.cache/radm/`, never volatile `/tmp` paths.

Commit only manifests, hashes, commands, sample ids, and regeneration metadata. Do not commit datasets, checkpoints, tensors, images, or large generated artifacts.

## Required S0–S4 order

Each stage must pass before the next begins. Record each stage in the tracking issue and `models/radm/TRAINING.md`.

### S0 — topology and static configuration

Add a named `test_s0_radm_topology` that asserts parameter-count equality, explicit state-dict coverage, same-seed same-input forward equality after copying original weights into the package model, schedule/derived-buffer equality at real dataset scale, optimizer/EMA/sampler/scheduler state, effective class count, label mapping, text-feature shape/mask behavior, and package-side execution using the package model itself.

A topology failure stops all later stages.

### S1 — fixed-batch pre-optimizer trace

With the same batch and captured RNG state, compare transformed inputs, boxes, text features/masks, timestep/noise, diffusion inputs, backbone/head outputs, auxiliary outputs, classification/L1/GIoU/total losses, and first-divergence diagnostics.

### S2 — one optimizer step

Compare gradients, clipping, optimizer state, post-step parameters, scheduler state, learning rate, and checkpoint state.

### S3 — deterministic multi-batch run

Verify repeated loader behavior, RNG consumption, scheduler cadence, logging/accumulation, clipping, checkpoint hooks, and evaluation/validation hooks that consume RNG.

### S4 — deterministic loader stream

Compare sample order, split membership, image transforms, annotation filtering, box conversion, category ids, text-feature lookup, masks/padding, and validation-stream behavior.

Post issue evidence after S0–S2 and again after S3–S4 with commands, artifact paths, hashes, tolerances, and first-divergence diagnostics.

## Explicit S5 gate

Do not launch, document, or claim S5 until S0–S4 evidence exists and is linked from the issue and `TRAINING.md`.

Before any S5 launch, run the protocol real-scale lockstep probe for at least 300 optimizer steps after every training-path change.

Only then may a future PR run full training/evaluation for every claimed dataset. Training-seed `n=3` per dataset is the target; evaluation-seed evidence must be labeled separately. The training PR remains draft until S5 is confirmed for every claimed dataset. Do not add `parity-verified` before coordinator rerun with `PARITY_REQUIRE=1`.

## Documentation and tests

Add `models/radm/TRAINING.md` from `docs/templates/TRAINING.template.md`, including dataset/config tables, scheduler/recipe notes, seed policy, machine-readable S0–S5 evidence, per-dataset results, durable artifact paths, and copy-pasteable training/reference/conversion/evaluation/`from_pretrained` commands.

Add unit tests for config parsing, dataset/collator behavior, losses, optimizer/scheduler wiring, and LightningCLI construction; topology tests; gated S0–S4 vendor-parity tests; training-checkpoint conversion tests; and local `from_pretrained` smoke tests. Maintain at least 90% normal-selection coverage.

Run the training documentation and evidence checkers. S5 claims must fail closed when S0–S4 evidence is absent.

## Non-goals

- No training code, training configs, CGL-v2 materialization, or S5 commands in PR #209.
- No Hub publication or converted-weight claim before license and checkpoint provenance are confirmed.
- No accepted parity claim from synthetic smoke tests.
- No full dataset download in CI or unit tests.
- No S5 execution before S0–S4 and the real-scale lockstep probe.

## Acceptance criteria

- The package training model is the runtime RADM model, not a simplified substitute.
- The original Detectron2 topology guard passes for every claimed dataset/config.
- The vendor-only reference adapter exposes effective original training behavior.
- LightningCLI YAMLs reproduce original optimizer, scheduler, loader, transform, RNG, and checkpoint behavior.
- S0, S1, S2, S3, and S4 evidence is recorded in order.
- `models/radm/TRAINING.md` passes canonical template and stage-evidence checkers.
- Converted training checkpoints load through the existing RADM `from_pretrained` path.
- S5 remains explicitly unrun until all preceding evidence is complete.
- The implementation PR remains draft until the S5 gate is satisfied.

## Dependencies and open decisions

Dependencies:

1. PR #209 must first be reduced to its agreed inference-only scope and merged.
2. The cross-model lazy-import contract in [issue #260](https://github.com/creative-graphic-design/design-generators/issues/260) must be finalized before the RADM training PR merges.
3. Shared `traingen`/`traingen-parity` APIs and the training protocol remain authoritative.
4. CGL/CGL-v2 asset provenance, text-feature availability, and RADM license/checkpoint terms must be confirmed.

Open decisions:

- Whether the future claim covers legacy CGL, CGL-v2, or both; each dataset needs independent status.
- Whether the original four-class effective configuration or five-class CGL vocabulary is the intended target.
- Which text-feature artifact/version is authoritative and whether the original missing-feature fallback is claimed.
- Exact S5 compute and seed budget after S0–S4 pass.
- Coordinate implementation order with [issue #260](https://github.com/creative-graphic-design/design-generators/issues/260) if its accepted contract changes before the RADM training PR.

## References

- RADM issue #35: https://github.com/creative-graphic-design/design-generators/issues/35
- RADM PR #209: https://github.com/creative-graphic-design/design-generators/pull/209
- Checklist #60: https://github.com/creative-graphic-design/design-generators/issues/60
- Training parity #148: https://github.com/creative-graphic-design/design-generators/issues/148
- Training protocol #212: https://github.com/creative-graphic-design/design-generators/issues/212
- Durable storage #249: https://github.com/creative-graphic-design/design-generators/issues/249
- TRAINING.md template #253: https://github.com/creative-graphic-design/design-generators/issues/253
- Shared lazy training imports #260: https://github.com/creative-graphic-design/design-generators/issues/260
- RADM paper: https://arxiv.org/abs/2306.09086
- Original RADM implementation: https://github.com/JD-GenX/RADM

## Handoff status

This is a read-only draft. Do not create the issue until `setup-gh-cgd` has been run and `creative-graphic-design-dev` is verified with target push permission, `project`, and `read:project` scopes.

**Comment by creative-graphic-design-dev** [comment]
RADM training evidence checkpoint: https://github.com/creative-graphic-design/design-generators/pull/269

Current head: https://github.com/creative-graphic-design/design-generators/tree/0b7541fa12783102849ba189a66e4f1708c29009/models/radm
Training status and staged evidence scope: https://github.com/creative-graphic-design/design-generators/blob/0b7541fa12783102849ba189a66e4f1708c29009/models/radm/TRAINING.md

S0-S2 are accepted on the supported V100 diagnostic runtime; two-layer S3 remains diagnostic-only, with synchronized first_divergence=None and natural step-2 drift recorded. Aggregate evidence SHA-256: b4498680c80c04eb193e7166e79553fa8d9ab2e93120b039e4ba291f9b60677d (.cache/radm/s3/two-layer-20260814-rerun.json, not committed). S4 investigation may continue; S5 and the real-scale 300-step lockstep remain stopped. Existing tolerances and stage contracts are unchanged.

**Comment by creative-graphic-design-dev** [comment]
## RADM S5 contract amendment: run-007 accepted preflight

This comment records the user-approved amendment to the RADM training-reproduction contract for issue #261.

### Run-007 evidence chain

- Initial package/source state was bitwise equal.
- The seeded record-then-compare run produced 300 records with `batch_order_equal=300/300` and `rng_equal=300/300`.
- Step 1 forward and loss surfaces were equal except `loss_giou_1`, with `max_abs=1.1920928955078125e-07`; step-1 preclip gradient `max_abs=8.106231689453125e-06`.
- The isolated FrozenBN, wrapped-convolution, and complete bottleneck backward A/B was bitwise equal, so the remaining deep-backward residual is documented accumulation-order noise rather than an identified production-op mismatch.
- The loss profile shows discrete jumps at steps 7 and 9 (>10x step-to-step under the recorded max-surface profile), consistent with ULP-seeded amplification through discrete assignment sensitivity; this is retained as an analysis of the evidence, not as a new tolerance.

Evidence: `.cache/radm/s5-preflight/run-007-lockstep-300.jsonl` SHA-256 `a5e039e00c9230e58a8222cc8d6ed430eb1be8f8ba053e9173013033d8a91995`; backward probe `.cache/radm/s5-preflight/run-007-step1-backward-retry-probe.json` SHA-256 `9310612248c6927a58c563b9b80f94c2ff0bb29536ab3d4ac3ce286d9175232c`.

### Amended S5 preflight contract

Run-007 is `PASS` under this amended preflight contract:

1. Bitwise initial state.
2. Bitwise RNG and batch alignment for all 300 records.
3. Bitwise step-1 forward/loss surfaces, allowing only the recorded floating-point dust.
4. Step-1 preclip gradient absolute error at or below the existing S2 `atol=5e-5`.
5. A documented numerical-chaos analysis for discrete assignment operators, including the isolated backward A/B and the observed jump profile.

For architectures with discrete assignment operators, the training-reproduction endpoint is not an exact per-step loss lockstep beyond the documented ULP-amplification horizon when the package and original use differently composed but forward-equivalent graphs. The S5 claim is instead carried by the CGL-only multi-seed package-versus-original full-run training and evaluation distribution comparison, with the accepted training-seed plan.
